spec/abstract/Exceptions_A.thy

```diff
@@ -49,7 +49,7 @@ translations
   (type) "'a f_monad" <= (type) "(fault + 'a) s_monad"
   (type) "'a se_monad" <= (type) "(syscall_error + 'a) s_monad"
   (type) "'a lf_monad" <= (type) "(lookup_failure + 'a) s_monad"
-  (type) "'a p_monad" <=(type) "(interrupt + 'a) s_monad"
+  (type) "'a p_monad" <=(type) "(unit + 'a) s_monad"
```

Comment:
I don't really know anything about preemption or about these monads, but doesn't this change remove the ability to associate an irq with an exemption? If we were wanting to increase the flexibility of the preemption monad to allow it to be associated with things other than an irq, then wouldn't we want to introduce a `preemption` type, have `interrupt` as one of the possibilities, and then, say for MCS, allow others as necessary?

Reply:
Well, so, the point is, we don’t need to be able to associate an irq with a preemption. Or, at least, that is what we have in the C code, as I understand it. In the C code, preemption is just a label that conveys the information that preemption occurred, and that is what the change here is trying to capture.

Reply:
Ah OK. I'm not sure why we bothered to have the `interrupt` type there in the first place, then. Maybe it's historical

Reply:
It's historical, yes. They way the kernel implemented preemption changed a few times. Initially it followed Haskell in looking up the IRQ number inside the preemption point and passing it out, but then the mechanism changed and it was easier to look it up again outside the preemption point. From that point on, the IRQ number was actually ignored in Haskell, which didn't break anything, so we didn't see immediately that it can/should be removed.

Since MCS now adds a path where there is no IRQ number to associate with the preemption, it makes sense to sync up with C.